File: public_key_pinning_ios/ios/Classes/MethodHandler.swift
//
//  MethodHandler.swift
//  public_key_pinning
//
//  Created by Kunnatam Na Takuatung on 6/3/2565 BE.
//

import Foundation

class MethodHandler {
  var pinner:SSLPinner?
  
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "init":
        initSession(call, result: result)
    case "request":
      requestTrust(call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
  
  private func initSession(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String:Any] else {
      return result(FlutterError(code: "MISSING_ARGUMENT", message: "Unable to obtain method's arguments.", details: nil))
    }
    
    guard let evaluations = args["evaluations"] as? [[String:String]] else {
      return result(FlutterError(code: "INVALID_PARAMETER", message: "Unable to obtain method's parameter of key 'evaluations'.", details: nil))
    }
    
    do {
        let json = try JSONSerialization.data(withJSONObject: evaluations)
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let decodedEvaluations = try decoder.decode([Evaluation].self, from: json)
        decodedEvaluations.forEach{print($0)}
        pinner = try SSLPinner.init(decodedEvaluations)
        result(nil)
    } catch {
      return result(FlutterError(code: "LIBRARY_ERROR", message: "Decode Evaluations list failed", details: nil))
    }
  }
  
  private func requestTrust(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String:Any] else {
      return result(FlutterError(code: "MISSING_ARGUMENT", message: "Unable to obtain method's arguments.", details: nil))
    }
    
    guard let url = args["url"] as? String else {
      return result(FlutterError(code: "INVALID_PARAMETER", message: "Unable to obtain method's parameter(s) of key 'url'.", details: nil))
    }
    
    let headers = args["headers"] as? [String:String]
    
    if let p = pinner {
      p.request(url: url, headers: headers) {
        (success) in
        result(success)
      }
    } else {
      return result(FlutterError(code: "LIBRARY_ERROR", message: "Pinner is not being initialized.", details: nil))
    }
  }
}

File: public_key_pinning_android/android/src/main/kotlin/com/klaa/public_key_pinning_android/MethodHandler.kt
package com.klaa.public_key_pinning_android

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

class MethodHandler {
    var pinner: SSLPinner? = null

    fun handle(call: MethodCall, result: Result) {
        when (call.method) {
            "init" -> {
                initSession(call, result)
            }
            "request" -> {
                requestTrust(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initSession(call: MethodCall, result: Result) {
        val args = call.arguments as? MutableMap<String, Any> ?: return result.error(
                "MISSING_ARGUMENT",
                "Unable to obtain method's arguments.", null)


        val evaluations = args["evaluations"] as? MutableSet<MutableMap<String, String>>? ?: return result.error(
            "INVALID_PARAMETER",
            "Unable to obtain method's parameter of key 'evaluations'.", null)

//         return result.success(SSLPinner.init(evs))
    }

    private fun requestTrust(call: MethodCall, result: Result) {
        val args = call.arguments as? MutableMap<String, Any> ?: return result.error(
            "MISSING_ARGUMENT",
            "Unable to obtain method's arguments.", null)

        val url = args["url"] as? String
        val headers = args["headers"] as? MutableMap<String, String>?
        if (url == null || headers == null) {
            return result.error(
                "INVALID_PARAMETER",
                "Unable to obtain method's parameter(s) of key 'url' and/or 'headers'.", null)
        }

        pinner?.request(url, headers) {
            if (it) {
                result.success(it)
            }
        }
    }
}